### PR TITLE
deprecate app-state

### PR DESCRIPTION
**Pull request by chrisnojima**
The goal of this PR is to simplify and unify how we deal with app state settings. Currently we have a kinda weird one-off app-state.desktop which persists data to its own custom json file. This causes us to have a bunch of custom code that shuttles data back and forth between the ipcs to manage this data. Instead of that we do this:

1. On boot the main node thread reads the gui config.json file. It needs to know some things before the renderer even exists.
2. Renderer starts up and reads this file through the rpc mechanism
3. Changes to prefs (in settings, things like sounds on notifications, native frame, etc) use redux and as a side effect we use the rpc to write it down (like our normal flows)
4. If the electron window moves, the main thread dispatches an action to the renderer (our usual flow) to update, which as a side effect uses the rpc to write it down

### CODE REVIEW

**Comment by chrisnojima** [comment]
- [ ] after release

File: shared/actions/json/config.json [hunk 1]
```diff
@@ -138,6 +126,13 @@
       "_description": "remote electron window wants props sent",
       "component": "string",
       "param": "string"
+    },
+    "updateWindowState": {
+      "_description": "main electron window wants to store its state",
+      "windowState": "Types.WindowState",
+    },
+    "setUseNativeFrame": {
```
Comment (chrisnojima) [inline shared/actions/json/config.json]: moved from settings

File: shared/actions/platform-specific/index.desktop.tsx [hunk 2]
```diff
@@ -35,9 +31,7 @@ export async function saveAttachmentToCameraRoll() {
 }
 
 const showMainWindow = () => {
-  const mw = getMainWindow()
-  mw && mw.show()
-  showDockIcon()
+  SafeElectron.getApp().emit('KBkeybase', '', {type: 'showMainWindow'})
```
Comment (chrisnojima) [inline shared/actions/platform-specific/index.desktop.tsx]: we just want it shown. main thread handles dock management

File: shared/actions/platform-specific/index.desktop.tsx [hunk 3]
```diff
@@ -68,28 +62,6 @@ export const getContentTypeFromURL = (
   req.end()
 }
 
-const writeElectronSettingsOpenAtLogin = (
```
Comment (chrisnojima) [inline shared/actions/platform-specific/index.desktop.tsx]: 🔥

File: shared/actions/platform-specific/index.desktop.tsx [hunk 4]
```diff
@@ -130,24 +102,6 @@ function* initializeInputMonitor(): Iterable<any> {
   }
 }
 
-// get this value from electron and update our store version
-function* initializeAppSettingsState(): Iterable<any> {
-  const getAppState = () =>
-    new Promise(resolve => {
-      SafeElectron.getApp().once(
```
Comment (chrisnojima) [inline shared/actions/platform-specific/index.desktop.tsx]: 🔥

File: shared/actions/platform-specific/index.desktop.tsx [hunk 5]
```diff
@@ -397,17 +422,19 @@ export function* platformConfigSaga(): Saga.SagaGenerator<any, any> {
   yield* Saga.chainAction2(ConfigGen.updateNow, updateNow)
   yield* Saga.chainAction2(ConfigGen.checkForUpdate, checkForUpdate)
   yield* Saga.chainAction2(ConfigGen.daemonHandshakeWait, sendKBServiceCheck)
-  yield* Saga.chainAction2(SettingsGen.onChangeUseNativeFrame, setUseNativeFrame)
+  yield* Saga.chainAction2(ConfigGen.setUseNativeFrame, saveUseNativeFrame)
```
Comment (chrisnojima) [inline shared/actions/platform-specific/index.desktop.tsx]: we write to the rpc when this changes

File: shared/actions/platform-specific/index.desktop.tsx [hunk 6]
```diff
@@ -397,17 +422,19 @@ export function* platformConfigSaga(): Saga.SagaGenerator<any, any> {
...
   yield* Saga.chainAction2(ConfigGen.loggedIn, initOsNetworkStatus)
+  yield* Saga.chainAction2(ConfigGen.updateWindowState, saveWindowState)
 
   if (isWindows) {
     yield* Saga.chainGenerator<ConfigGen.DaemonHandshakePayload>(ConfigGen.daemonHandshake, checkRPCOwnership)
   }
 
   yield Saga.spawn(initializeUseNativeFrame)
+  yield Saga.spawn(initializeNotifySound)
```
Comment (chrisnojima) [inline shared/actions/platform-specific/index.desktop.tsx]: we init from the rpc on start

File: shared/app/app-state.d.ts [hunk 7]
```diff
@@ -1,21 +0,0 @@
-export type State = {
```
Comment (chrisnojima) [inline shared/app/app-state.d.ts]: killing this file is essentially the point of this whole thing

File: shared/app/app-state.d.ts [hunk 8]
```diff
@@ -1,21 +0,0 @@
...
-  x: number | null
-  y: number | null
-  width: number
-  height: number
-  windowHidden: boolean
-  isMaximized: boolean | null
-  isFullScreen: boolean | null
-  displayBounds: any | null
-  tab: string | null
```
Comment (chrisnojima) [inline shared/app/app-state.d.ts]: never actually used

File: shared/desktop/app/dock-icon.desktop.tsx [hunk 9]
```diff
@@ -1,22 +0,0 @@
-import * as SafeElectron from '../../util/safe-electron.desktop'
-
-const changeDock = (show: boolean) => {
```
Comment (chrisnojima) [inline shared/desktop/app/dock-icon.desktop.tsx]: having a standalone file for this tiny function that's only imported from one place hurts more than it helps

File: shared/desktop/app/menu-bar.desktop.tsx [hunk 10]
```diff
@@ -6,12 +6,19 @@ import {resolveImage, resolveRootAsURL} from './resolve-root.desktop'
 import {showDevTools, skipSecondaryDevtools} from '../../local-debug.desktop'
 import logger from '../../logger'
 
-const htmlFile = resolveRootAsURL('dist', `menubar${__DEV__ ? '.dev' : ''}.html`)
+const htmlFile = resolveRootAsURL('dist', `menubar${__DEV__ ? '.dev' : ''}.html?param=`)
```
Comment (chrisnojima) [inline shared/desktop/app/menu-bar.desktop.tsx]: so the url scheme matches what we do elsewhere. the main thread finds things from url now

File: shared/desktop/app/menu-helper.desktop.tsx [hunk 11]
```diff
@@ -1,16 +1,14 @@
...
 const windowQuit = () => {
-  SafeElectron.getApp().emit('close-windows')
-  hideDockIcon()
+  Electron.app.emit('KBkeybase', '', {type: 'closeWindows'})
 }
 
-export default function makeMenu(window: any) {
-  const editMenu = {
+export default function makeMenu(window: Electron.BrowserWindow) {
+  const editMenu = new Electron.MenuItem({
```
Comment (chrisnojima) [inline shared/desktop/app/menu-helper.desktop.tsx]: the types that ship with electron only allow partial menu definitions if using the constructor. else you have to pass every value in, and some types are wider than they should be (see `role`), so we just use the constructor version

File: shared/desktop/app/menu-helper.desktop.tsx [hunk 12]
```diff
@@ -21,58 +19,58 @@ export default function makeMenu(window: any) {
...
-      {label: 'Reset zoom ', role: 'resetZoom'},
-      {label: 'Minimize', role: 'minimize'},
-      {accelerator: 'CmdOrCtrl+W', label: 'Close', role: 'close'},
-      {type: 'separator'},
-      {label: 'Bring All to Front', role: 'front'},
-    ].concat(
-      __DEV__
+    submenu: Electron.Menu.buildFromTemplate([
+      new Electron.MenuItem({accelerator: 'CmdOrCtrl+=', label: 'Zoom In', role: 'zoomin'}),
+      new Electron.MenuItem({label: 'Zoom Out', role: 'zoomout'}),
```
Comment (chrisnojima) [inline shared/desktop/app/menu-helper.desktop.tsx]: note the case error previously! `zoomout !== zoomOut`

File: shared/desktop/app/node.desktop.tsx [hunk 13]
```diff
@@ -211,9 +244,68 @@ const plumbEvents = () => {
           startWinService()
         }
         break
-      case 'requestShowDockIcon':
+      case 'closeWindows': {
```
Comment (chrisnojima) [inline shared/desktop/app/node.desktop.tsx]: vs before, the main renderer would make the other renderers, but really we want the node thread to do this work.

File: shared/desktop/app/node.desktop.tsx [hunk 14]
```diff
@@ -211,9 +244,68 @@ const plumbEvents = () => {
...
+
+        break
+      }
+      case 'closeRenderer': {
+        const w = findRemoteComponent(action.payload.windowComponent, action.payload.windowParam)
+        w && w.close()
         break
+      }
+      case 'rendererNewProps': {
+        const w = findRemoteComponent(action.payload.windowComponent, action.payload.windowParam)
```
Comment (chrisnojima) [inline shared/desktop/app/node.desktop.tsx]: shuttling the props between the windows is now the role of the main node thread

File: shared/desktop/app/window.desktop.tsx [hunk 15]
```diff
@@ -1,67 +0,0 @@
-import {showDockIcon, hideDockIcon} from './dock-icon.desktop'
-import menuHelper from './menu-helper.desktop'
-import * as SafeElectron from '../../util/safe-electron.desktop'
-
-export default class Window {
```
Comment (chrisnojima) [inline shared/desktop/app/window.desktop.tsx]: this is only used by main window and i've always found it confusing

File: shared/desktop/remote/component-loader.desktop.tsx [hunk 16]
```diff
@@ -7,15 +7,14 @@ import ReactDOM from 'react-dom'
...
 import {disable as disableDragDrop} from '../../util/drag-drop'
-import {setupContextMenu} from '../app/menu-helper.desktop'
 import ErrorBoundary from '../../common-adapters/error-boundary'
 import {initDesktopStyles} from '../../styles/index.desktop'
 
 disableDragDrop()
 
 module.hot && module.hot.accept()
 
-type RemoteComponents = 'unlock-folders' | 'menubar' | 'pinentry' | 'tracker' | 'tracker2'
```
Comment (chrisnojima) [inline shared/desktop/remote/component-loader.desktop.tsx]: we killed the old tracker awhile ago

File: shared/desktop/remote/sync-avatar-props.desktop.tsx [hunk 17]
```diff
@@ -2,14 +2,12 @@
...
 import * as I from 'immutable'
 import * as React from 'react'
-import * as SafeElectron from '../../util/safe-electron.desktop'
 import * as Container from '../../util/container'
 import {memoize} from '../../util/memoize'
 
 type OwnProps = {
   usernames: I.Set<string>
   setUsernames: (arg0: I.Set<string>) => void
-  remoteWindow: SafeElectron.BrowserWindowType | null
```
Comment (chrisnojima) [inline shared/desktop/remote/sync-avatar-props.desktop.tsx]: all these remote window hocs used to actually have a direct instance of the browerwindow. instead we just use the `windowComponent` and `windowParam` as a unique identifier and delegate to the node process

File: shared/desktop/remote/sync-browser-window.desktop.tsx [hunk 18]
```diff
@@ -13,119 +10,35 @@ type Props = {
...
-        return
-      }
-      const webContents = this._remoteWindow.webContents
-      if (showDevTools && !skipSecondaryDevtools) {
-        webContents.openDevTools({mode: 'detach'})
+  const RemoteWindowComponent = (props: Props) => {
+    React.useEffect(() => {
+      const {windowOpts, windowComponent, windowParam, windowPositionBottomRight} = props
+
+      SafeElectron.getApp().emit('KBkeybase', '', {
```
Comment (chrisnojima) [inline shared/desktop/remote/sync-browser-window.desktop.tsx]: we ask the node thread to make and destroy this window when this component is mounted/unmounted (vs doing it ourselves before)

File: shared/desktop/renderer/main2.desktop.tsx [hunk 19]
```diff
@@ -12,7 +12,6 @@ import configureStore from '../../store/configure-store'
 import * as SafeElectron from '../../util/safe-electron.desktop'
 import {makeEngine} from '../../engine'
 import {disable as disableDragDrop} from '../../util/drag-drop'
-import {setupContextMenu} from '../app/menu-helper.desktop'
```
Comment (chrisnojima) [inline shared/desktop/renderer/main2.desktop.tsx]: the menuHelper default export actually does this itself

File: shared/pinentry/index.desktop.tsx [hunk 20]
```diff
@@ -82,10 +82,9 @@ class Pinentry extends React.Component<Props, State> {
           ...globalStyles.flexBoxColumn,
           backgroundColor: globalColors.white,
           paddingBottom: globalMargins.medium,
-          paddingTop: globalMargins.medium,
```
Comment (chrisnojima) [inline shared/pinentry/index.desktop.tsx]: this is a random fix that fixes an issue @mlsteele found. i was already in here so...

File: shared/pinentry/index.desktop.tsx [hunk 21]
```diff
@@ -82,10 +82,9 @@ class Pinentry extends React.Component<Props, State> {
...
         }}
       >
-        <Header icon={true} title="" onClose={this.props.onCancel} />
+        <Header icon={false} title="" onClose={this.props.onCancel} windowDragging={true} />
```
Comment (chrisnojima) [inline shared/pinentry/index.desktop.tsx]: icon off as it overlaps w/ the system close stuff

**Comment by chrisnojima** [comment]
- [ ] after release
- [ ] test linux
- [ ] test windows
- [ ] mobile

**Comment by chrisnojima** [comment]
@keybase/react-hackers this'll go in after the release. we can review it after. will need to test linux/windows 
cc: @heronhaye

File: shared/desktop/app/main-window.desktop.tsx [hunk 22]
```diff
@@ -32,75 +35,166 @@ export default function() {
...
+
+let useNativeFrame = defaultUseNativeFrame
+
+const loadWindowState = () => {
+  const filename = dataRoot + 'gui_config.json'
 
-  const openedAtLogin = app.getLoginItemSettings().wasOpenedAtLogin
+  let openAtLogin = true
+  try {
+    const s = fs.readFileSync(filename, {encoding: 'utf8'})
```
Comment (heronhaye) [inline shared/desktop/app/main-window.desktop.tsx]: how will this work with sandboxing?
also, does this pr migrate your old setting?

**Pull request merged**